### DISCUSSION TITLE
MVVM helper/base classes from TinyMvvm

### DISCUSSION THREAD

**Discussion by dhindrik** [CommunityToolkit/Maui#990] [New Feature Discussions]
In TinyMvvm I have some helper methods and properties in a ViewModel base class that makes life a bit easier when doing Mvvm. Compared with the first versions of TinyMvvm I have stripped away a lot of stuff for the MAUI version.  But instead of having this code in a separate library it would be great to add it to the community toolkit so more devs can use it easier and to reuse the amount of dependencies. Maybe it should be a separate package because it has dependencies to CommuniytToolkit.Mvvm. Even if it is Mvvm it should be in the MAUI Community toolkit, because the code is specific to MAUI.

To get an ide about the implementation, you can look here. The implementations will be cleaner because I don't think everything should there should be included.

**New classes/interfaces**
* IViewModelBase
* ViewModelBase
* ViewBase 

ViewBase will not have any new properties or methods, just a few overrides that trigger methods in the ViewModel.

**IViewModelBase**
```csharp
     /// <summary>
    /// This method will run after that the ViewModel has been set as BindingContext of the View
    /// </summary>
    Task Initialize();

    /// <summary>
    /// This method runs when navigation parameters are set.
    /// </summary>
    /// <returns></returns>
    Task OnParameterSet();

    /// <summary>
    /// This method will run every time the view is appearing.
    /// </summary>
    Task OnAppearing();

    /// <summary>
    /// This method will run every time the view is disappearing.
    /// </summary>
    Task OnDisappearing();

    /// <summary>
    /// Parameter that passed to this ViewModel when navigated, will not be available in the constructor, use it when Initialize is running or later. 
    /// </summary>
   IDictionary<string, object>? NavigationParameters { get; set; }

    /// <summary>
    /// Queryparameters as a Dictionary.
    /// </summary>
    IDictionary<string, string>? QueryParameters { get; set; }
```

TinyMvvm also have methods for OnApplicationSleep and OnApplicationResume, but that requires a custom Application class, even if it nice I don't know if it is the right way to do it in the community toolkit.

**Comment by pictos** [comment]
Hey @dhindrik, thanks for your suggestion. IMHO, this looks like something that would be good to have in an MVVM framework, and (if nothing changed from the last time the team discuss it) we would like to avoid our toolkit following that path.
Anyway I'll rise up this with the team and double-check (:

**Comment by dersia** [comment]
I think this is a great idea. Although, I agree, this should be in the mvvm package, I see a bug issue with that, because the implemtation of this would be very platform specific.

Methods like `OnAppearing()` and `OnDisappearing()` need to be implemented differently in Mauiverse vs in WinUIverse, because they depend on either the page or the DI/NavigationService. So the only Platform abstraction that could be done in the mvvm toolkit would be the interface and the the maui toolkit has to take a dependency on the mvvm toolkit to implement it.

So I guess it would make much mor sense to implent it in this toolkit. There is also a issue for Source Generators for BindableObjects, which is also very specific to mvvm and I guess that one could even be better in the mvvm toolkit (although not quite sure, I have to look into the proposal again).

Today I was actually looking for a feature like the one proposed in this discussion and unfortunately there is nothing like this in any of the toolkits, which means I have to use one of the other mvvm toolkits like prism, mvvmcross or tinymvvm to get something that small just to be able to do work on appearing.

So I would support this feature and I'd love to see it in the toolkit, for what it's worth.

**Comment by CliffAgius** [comment]
Closed as per the Community Stand-up discussion here: https://youtu.be/N9wMcBP4jtg?t=2889
